compiler/testData/diagnostics/tests/controlFlowAnalysis/assignedInTryWithCatch.kt

```diff
@@ -0,0 +1,70 @@
+fun assignedInTry() {
+    val a: Int
+    try {
+        a = 42
+    } catch (e: Exception) {
+    } finally {
+    }
+    <!UNINITIALIZED_VARIABLE!>a<!>.hashCode()
+}
+
+fun sideEffectBeforeAssignmentInTry(s: Any) {
+    val a: Int
+    try {
+        s as String // Potential cast exception
+        a = 42
+    } catch (e: Exception) {
+    } finally {
+    }
+    <!UNINITIALIZED_VARIABLE!>a<!>.hashCode()
+}
+
+fun assignedInTryAndCatch() {
+    val a: Int
+    try {
+        a = 42
+    } catch (e: Exception) {
+        <!VAL_REASSIGNMENT!>a<!> = 41
+    } finally {
+    }
+    a.hashCode()
+}
+
+fun sideEffectBeforeAssignedInTryAndCatch(s: Any) {
+    val a: Int
+    try {
+        s as String // Potential cast exception
+        a = 42
+    } catch (e: Exception) {
+        s as String // Potential cast exception
+        <!VAL_REASSIGNMENT!>a<!> = 41
+    } finally {
+    }
+    a.hashCode()
```

Comment:
It will be good add here `s.length` call, so we can check behavior smartcast from finally blocks also (or create separate test for it)

Reply:
It would be better to put assignment-related tests together, and separate smartcasts-related tests. I found at https://github.com/JetBrains/kotlin/pull/3776 that many existing tests do that way.

To that end, I tried to add similar smartcast tests, and found that... FIR DFA doesn't behave same as old FE, even on master. At a glance, type info at FIR DFA looks much more aggregated form, and once type _approved_, one variable's changed type is used/retrieved everywhere. See https://github.com/JetBrains/kotlin/pull/3843. I think it's a different issue. Even with these changes, smartcast info is not changed. I'll keep learning DFA and investigating those tests in that PR.

Reply:
As mentioned internally, I think the way FIR DFA handles try-catch-finally needs some improvement. In particular, around here: https://github.com/JetBrains/kotlin/blob/master/compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/dfa/FirDataFlowAnalyzer.kt#L656, different paths need _fork_ed flows; otherwise, all nodes inside try-catch-finally use the same persistent flow, resulting in incorrect type info at the end. Again, I believe this is a separate issue, and can be tracked at that PR. Though, I need this PR to be merged first, since it will depend on CFG changes in this PR